### PR TITLE
First steps toward documentation

### PR DESCRIPTION
**Pull request by nils-werner**
Navigating this library without any docs is a bit tedious... so I thought I could fix that by starting to document the API.

If you are OK with me doing this I can continue pushing more and more docs commits into this PR until every function is properly documented.

For now you can generate the docs using

    pip install -e .[docs]
    python setup.py build_sphinx

and then opening `build/sphinx/html/index.html`.

To have a publicly accessible docs page you could register and set up a project at https://readthedocs.org/

### CODE REVIEW

**Comment by hameerabbasi** [comment]
This is very welcome! It will help gain wider adoption and, in turn, more development. Keep in mind, though, the library is fast-changing and you may require quite a bit of rebasing,

**Comment by hameerabbasi** [comment]
@nils-werner Are you still working on this? It'd be great to get this in before `0.2`.

**Comment by hameerabbasi** [comment]
cc @mrocklin What do you think of merging this right away?

**Comment by nils-werner** [comment]
Yes, I am still working on it. Things just slowed down a bit for the holidays :-)

I'd be happy to do an initial API reference for `0.2`. I guess additional docs like usage examples could follow once all rudimentary docstrings are in place.

**Comment by hameerabbasi** [comment]
The API should hopefully be pretty stable now, just new things being added. For some functions, it'd be nice if we linked to the relevant page in the Numpy/Scipy docs.

**Comment by hameerabbasi** [comment]
There! I've done a lot of the work for you! Hopefully this looks fine in the generated HTML, I'm no expert in reStructuredText.

**Comment by mrocklin** [comment]
The commit history here is a bit odd.  I think that this needs to be rebased on top of a clean master branch

**Comment by hameerabbasi** [comment]
This should fix it.

**Comment by mrocklin** [comment]
Some comments:

1.  Thank you for setting up the docstrings.  They're great.  
2.  I'm not sure what to do about the copyright name.  Putting my name there probably isn't correct given that other people (you) are writing most of the documentation so far.  Perhaps "Sparse Developers" ?
3.  When I set up sphinx I usually use the sphinx quickstart, which sets up things like `Makefile`s.  It might be worth considering this instead
4.  This will need a home somewhere.  The name `sparse` is taken on readthedocs.  Any suggestions for another name?

**Comment by hameerabbasi** [comment]
Here are my responses, but I invite @nils-werner to chime in:

1. Thanks!
2. "Sparse Developers" sounds fine, given changing the copyright to Numfocus or Dask as per #7 won't be an issue.
3. I believe that's what @nils-werner has done, or maybe something similar. Maybe he can comment?
4. Maybe `python-sparse`? (Another project already exists with the name `pysparse` and I don't want to cause confusion. Also, `sparsepy` sounds like a knock-off and is harder to say.)

**Comment by nils-werner** [comment]
> When I set up sphinx I usually use the sphinx quickstart,

Yes, I have sphinx quickstart a while back, but realized that `Makefile` and `make.bat` are pretty redundant and all you need is the `conf.py`. Nowadays I only copy and adapt the config file :-)

>  The name sparse is taken on readthedocs. Any suggestions for another name?

maybe some variation of `ndsparse`? I also contacted the owner of sparse.readthedocs, maybe he will agree to change the name...

**Comment by hameerabbasi** [comment]
@nils-werner, is it possible to revert nils-werner/sparse@cfd7c9bb68213557b9b37ff6d6fe73c054168b3f? I ask because I find it quite convenient and useful to just use `make html`, and judging from @mrocklin's comment, he does, too. It doesn't hurt anyone to have it in there. :)

**Comment by nils-werner** [comment]
Really? I much prefer less boilerplate and doing `python setup.py build_sphinx`

**Comment by mrocklin** [comment]
No strong preference from me.  I do find the `make html` approach more familiar and common, but I don't care that much.  

We would want to make sure that whatever choice we make is compatible with readthedocs.

**Comment by hameerabbasi** [comment]
@nils-werner Is it possible to add some structure to the TOC and API reference with links, either with `numpydoc` or otherwise? Unfortunately the best approaches I could find are to simply list everything out manually, which I really don't want to do.

**Comment by hameerabbasi** [comment]
This generates separate pages for API docs for every function, class and attribute. For now, I got it working by going to `sphinx/ext/autosummary/generate.py` and setting `imported_members=True` in the function `generate_autosummary_docs`. I've opened sphinx-doc/sphinx#4372 for our case, maybe there's already an option and I'm not seeing it.

**Comment by hameerabbasi** [comment]
When trying to convert `coords` and `data` into properties and properly document them, I ran into sphinx-doc/sphinx#4379.

**Comment by hameerabbasi** [comment]
cc @mrocklin @nils-werner I think this is ready for prime time. Please give it a look.

I added all the generated autosummaries to bypass needing to change the Sphinx codebase. Looks like that's how Numpy does it ([see this directory for example](https://github.com/numpy/numpy/tree/master/doc/source/reference)).

@nils-werner Did you hear back about the name change? If not, we should decide a RTD name and go with it.

@mrocklin I think we should move forward with releasing v0.2 once the docs are up.

File: docs/_templates/autosummary/base.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{{ objname | escape | underline}}
+
+.. currentmodule:: {{ module }}
+
+.. auto{{ objtype }}:: {{ objname }}
```
Comment (mrocklin) [inline docs/_templates/autosummary/base.rst]: Do we need these or are standard solutions ok?

File: docs/generated/sparse.COO.abs.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+COO\.abs
+========
+
+.. currentmodule:: sparse
+
+.. automethod:: COO.abs
```
Comment (mrocklin) [inline docs/generated/sparse.COO.abs.rst]: Is there an easy way to auto-generate these files as we change the API?

File: docs/generated/sparse.COO.rst [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+.. autoclass:: COO
+
+   
+   
+   .. rubric:: Attributes
+   .. autosummary::
+      :toctree:
+   
+      COO.T
+
```
Comment (mrocklin) [inline docs/generated/sparse.COO.rst]: Can we remove extra endlines here?

File: sparse/core.py [hunk 4]
```diff
@@ -362,6 +530,69 @@ def _reduce(method, *args, **kwargs):
...
+            If reducing an all-zero axis would produce a nonzero result.
+
+        Notes
+        -----
+        This function internally calls :obj:`COO.sum_duplicates` to bring the array into
+        canonical form.
+
+        See Also
+        --------
+        numpy.ufunc.reduce : A similar Numpy method.
```
Comment (mrocklin) [inline sparse/core.py]: If you're feeling energetic you might also link to these with intersphinx.  Here is a recent PR to dask that does this: https://github.com/dask/dask/pull/3050

File: docs/generated/sparse.COO.abs.rst (hunk 2 above)
Comment (hameerabbasi) [reply docs/generated/sparse.COO.abs.rst]: Yes, if we change the API, we just need to delete the `docs/generated` directory and re-run Sphinx with a certain modification. See https://github.com/mrocklin/sparse/pull/43#issuecomment-355033794 for details. However, for now, as RTD will have a standard, unmodified version of Sphinx, we need to include them. Once sphinx-doc/sphinx#4372 is fixed, we don't need to include them in the repository at all, they will be autogenerated.

File: docs/_templates/autosummary/base.rst (hunk 1 above)
Comment (hameerabbasi) [reply docs/_templates/autosummary/base.rst]: I could try removing them. However, we at least need `module.rst`. The rest, I'm not sure of.

File: docs/generated/sparse.COO.rst (hunk 3 above)
Comment (hameerabbasi) [reply docs/generated/sparse.COO.rst]: These are auto-generated. It's either ugliness in here or in the files that generate them.

File: sparse/core.py (hunk 4 above)
Comment (hameerabbasi) [reply sparse/core.py]: They already are linked. Well, most of them, anyway.

Comment (hameerabbasi) [reply sparse/core.py]: It seems it auto-detects most of them. The ones it doesn't, we need to just add `:obj:` to.

**Comment by nils-werner** [comment]
I would love if `COO` classmethods and regular methods would be separated, but it doesn't seem like Sphinx supports that...

**Comment by hameerabbasi** [comment]
Yes. For sometime in the future, we might want to manually write `sparse.COO.rst` and group methods into utility functions, reductions, class methods, and whatever else, with details for each part. If you're up to the task, it would be quite welcome.

We _will_ need to modify the auto-generated files at some point in the future.

**Comment by mrocklin** [comment]
It might be nice to have the COO class's API on the same level as the top-level functions.  People have to click down a few levels to get to these, but they're probably as important.

Typically I move API documentation to an `api.html` page and then have some front-matter explaining the project on the main page.  Happy either way though.

File: docs/generated/sparse.COO.rst (hunk 3 above)
Comment (mrocklin) [reply docs/generated/sparse.COO.rst]: OK, if they're auto-generated then I remove my comment

File: docs/generated/sparse.COO.abs.rst (hunk 2 above)
Comment (mrocklin) [reply docs/generated/sparse.COO.abs.rst]: OK.  Perhaps we should have a README or Makefile in the docs directory that explains these steps?  Will this work well with readthedocs?

File: sparse/core.py (hunk 4 above)
Comment (mrocklin) [reply sparse/core.py]: Ah, alright.  Glad to hear it

**Comment by mrocklin** [comment]
I'm also happy to just merge if you'd like to get this in and release.

**Comment by mrocklin** [comment]
This is clearly an improvement :)

File: docs/generated/sparse.COO.abs.rst (hunk 2 above)
Comment (hameerabbasi) [reply docs/generated/sparse.COO.abs.rst]: A Makefile is overdo, as we might need to manually modify these files at some point and if the makefile removes them it's no good. We'll probably add howto's for contributions and docs at some point but right now I think it's too early for that.

If these files already exist in the repository, then the _unmodified_ version of Sphinx picks them up and builds the docs nicely. This is actually _why_ I've added them in here. It's just auto-generating these files (specifically `generated/sparse.rst`) that's the issue.

**Comment by mrocklin** [comment]
Ah I see.  Sounds good to me

**Comment by hameerabbasi** [comment]
I'll probably add some front matter and segment out `generated/sparse.COO.rst` today and tomorrow.

As for adding the `COO` class API on the same level... It might actually make things too cluttered (too long of a list) on one page, and it wouldn't maintain consistency. It might be worth it to add links to the `COO` object in the front matter, or maybe on the "Getting Started" page.

The "too cluttered" argument also applies to the `COO` page itself, which is why I'm taking on segmenting it.

I was thinking of something like the following structure (feel free to propose changes):

- Introduction
  - Motivation
- Getting Started
  - Constructing `COO` objects
    - From `coords`/`data`
    - From Numpy arrays
    - From Scipy Sparse matrices
    - Random `COO` objects
  - Operations on `COO` objects
    - Basic operations (describe operators here)
    - Element-wise operations (describe `ufunc`s here)
    - Reductions
    - Advanced operations (`dot`, `tensordot`, `stack`)
- API Reference
    - Everything we already have here.

**Comment by hameerabbasi** [comment]
cc @mrocklin @nils-werner Any further input is welcome.

**Comment by hameerabbasi** [comment]
@nils-werner I segmented out `sparse.COO.rst` a bit. Is this closer to what you wanted?

@mrocklin I'm trying to figure out a way to make the `sparse` module the head of the API docs. The number of clicks might be reduced that way.

**Comment by nils-werner** [comment]
> I segmented out sparse.COO.rst a bit. Is this closer to what you wanted?

Great!

**Comment by hameerabbasi** [comment]
@nils-werner @mrocklin Any further input would be appreciated! Otherwise, I think this is ready to merge.

Keep in mind that we can update docs as we go. No need to keep it in a separate branch unless it's really needed.

**Comment by mrocklin** [comment]
+1 from me.  Feel free to merge.  Things that might be nice to see in the future:

1.  Instructions on how to build docs.  The current `python setup.py build_sphinx` may be standard, but I personally hadn't seen it before.  I had to check the comments in this PR to learn how to build things
2.  Verification that this will work with readthedocs.  Hopefully NumFOCUS lets us use the sparse.pydata.org name

**Comment by hameerabbasi** [comment]
`sphinx-build ./docs/ ./build/` also works, in case you were wondering. Also, it works with the [command in ReadTheDocs](https://docs.readthedocs.io/en/latest/builds.html), which is `sphinx-build -b html . _build/html`, run from the `docs` directory, so it does work with RTD. 

I'll add a `CONTRIBUTING` in a separate PR.

**Pull request merged**